internal/registry/registry.go

```diff
@@ -0,0 +1,241 @@
+package registry
+
+import (
+	"archive/tar"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"runtime"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/crane"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/google/go-containerregistry/pkg/v1/remote/transport"
+	ootov1alpha1 "github.com/qbarrand/oot-operator/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+const (
+	driverToolkitJSONFile = "etc/driver-toolkit-release.json"
+)
+
+type DriverToolkitEntry struct {
+	ImageURL            string `json:"imageURL"`
+	KernelFullVersion   string `json:"kernelFullVersion"`
+	RTKernelFullVersion string `json:"RTKernelFullVersion"`
+	OSVersion           string `json:"OSVersion"`
+}
+
+// [TODO] make this struct public, once there is a code that is going to use it
+type repoPullConfig struct {
+	repo        string
+	authOptions []crane.Option
+}
+
+//go:generate mockgen -source=registry.go -package=registry -destination=mock_registry_api.go
+
+type Registry interface {
+	ImageExists(ctx context.Context, image string, po ootov1alpha1.PullOptions) (bool, error)
+	ExtractToolkitRelease(v1.Layer) (*DriverToolkitEntry, error)
+	GetLayersDigests(ctx context.Context, image string) ([]string, *repoPullConfig, error)
+	GetLayerByDigest(digest string, pullConfig *repoPullConfig) (v1.Layer, error)
+}
+
+type registry struct{}
+
+func NewRegistry() Registry {
+	return &registry{}
+}
+
+func (r *registry) ImageExists(ctx context.Context, image string, po ootov1alpha1.PullOptions) (bool, error) {
+	_, _, err := r.getImageManifest(ctx, image, &po)
+	if err != nil {
+		te := &transport.Error{}
+		if errors.As(err, &te) && te.StatusCode == http.StatusNotFound {
+			return false, nil
+		}
+		return false, fmt.Errorf("could not get image %s: %w", image, err)
+	}
+	return true, nil
+}
+
+func (r *registry) GetLayersDigests(ctx context.Context, image string) ([]string, *repoPullConfig, error) {
+	manifest, pullConfig, err := r.getImageManifest(ctx, image, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get manifest from image %s: %w", image, err)
+	}
+
+	digests, err := r.getLayersDigestsFromManifestStream(manifest)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get layers digests from manifest of the image %s: %w", image, err)
+	}
+
+	return digests, pullConfig, nil
+}
+
+func (r *registry) GetLayerByDigest(digest string, pullConfig *repoPullConfig) (v1.Layer, error) {
+	return crane.PullLayer(pullConfig.repo+"@"+digest, pullConfig.authOptions...)
+}
+
+func (r *registry) ExtractToolkitRelease(layer v1.Layer) (*DriverToolkitEntry, error) {
```

Comment:
I would keep that only for the downstream version.

Reply:
Why? even in upstream we will define that image must contain drtvier toolkit release file in order for preflight to work. Maybe the name of the file(and function) should be changed, but the meaning stays the same